tests/integration-tests/src/commonTest/kotlin/test/WatcherErrorHandlingTest.kt

```diff
@@ -0,0 +1,274 @@
+import com.apollographql.apollo3.cache.normalized.api.MemoryCacheFactory
+import com.apollographql.apollo3.cache.normalized.fetchPolicy
+import com.apollographql.apollo3.cache.normalized.refetchPolicy
+import com.apollographql.apollo3.cache.normalized.store
+import com.apollographql.apollo3.cache.normalized.watch
+import com.apollographql.apollo3.exception.ApolloCompositeException
+import com.apollographql.apollo3.exception.ApolloHttpException
+import com.apollographql.apollo3.exception.CacheMissException
+import com.apollographql.apollo3.integration.normalizer.EpisodeHeroNameQuery
+import com.apollographql.apollo3.integration.normalizer.type.Episode
+import com.apollographql.apollo3.mockserver.MockServer
+import com.apollographql.apollo3.mockserver.enqueue
+import com.apollographql.apollo3.testing.receiveOrTimeout
+import com.apollographql.apollo3.testing.runTest
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import testFixtureToUtf8
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import kotlin.test.assertIs
+
+@OptIn(ApolloExperimental::class)
+class WatcherErrorHandlingTest {
+  private lateinit var mockServer: MockServer
+  private lateinit var apolloClient: ApolloClient
+  private lateinit var store: ApolloStore
+
+  private suspend fun setUp() {
+    store = ApolloStore(MemoryCacheFactory(), cacheKeyGenerator = IdCacheKeyGenerator)
+    mockServer = MockServer()
+    apolloClient = ApolloClient.Builder().serverUrl(mockServer.url()).store(store).build()
+  }
+
+  private suspend fun tearDown() {
+    mockServer.stop()
+  }
+
+  @Test
+  fun fetchIgnoreAllErrors() = runTest(before = { setUp() }, after = { tearDown() }) {
+    val channel = Channel<EpisodeHeroNameQuery.Data?>()
+    val jobs = mutableListOf<Job>()
+
+    jobs += launch {
+      apolloClient.query(EpisodeHeroNameQuery(Episode.EMPIRE))
+          .fetchPolicy(FetchPolicy.CacheFirst)
+          .watch()
+          .collect {
+            channel.send(it.data)
+          }
+    }
+
+    jobs += launch {
+      apolloClient.query(EpisodeHeroNameQuery(Episode.EMPIRE))
+          .fetchPolicy(FetchPolicy.CacheOnly)
+          .watch()
+          .collect {
+            channel.send(it.data)
+          }
+    }
+
+    jobs += launch {
+      apolloClient.query(EpisodeHeroNameQuery(Episode.EMPIRE))
+          .fetchPolicy(FetchPolicy.NetworkFirst)
+          .watch()
+          .collect {
+            channel.send(it.data)
+          }
+    }
+
+    jobs += launch {
+      apolloClient.query(EpisodeHeroNameQuery(Episode.EMPIRE))
+          .fetchPolicy(FetchPolicy.NetworkOnly)
+          .watch()
+          .collect {
+            channel.send(it.data)
+          }
+    }
+
+    channel.assertEmpty()
+    jobs.forEach { it.cancel() }
```

Comment:
Doesn't this cancels before anything could happen? I would expect this test to throw an uncaught exception in the MockServer thread because to response is enqueued.

Reply:
I missed that `assertEmpty()` had a timeout. Then I guess it's ok. I still can't explain why the mockserver doesn't throw though?

Reply:
I changed the behavior when working on MockServer, now it handles the case by [returning 500](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/apollo-mockserver/src/jvmMain/kotlin/com/apollographql/apollo3/mockserver/MockServer.kt#L45) instead of crashing.

Reply:
Ah, I see. But [native is still crashing](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/apollo-mockserver/src/appleMain/kotlin/com/apollographql/apollo3/mockserver/QueueMockServerHandler.kt#L15).

Overall, I like the behaviour of crashing better. IIRC, this is what OkHttp MockServer does so it might be less surprising. Also it's more explicit?

Reply:
For native it's handled [here](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/apollo-mockserver/src/appleMain/kotlin/com/apollographql/apollo3/mockserver/Socket.kt#L137) actually! ([here](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/apollo-mockserver/src/jsMain/kotlin/com/apollographql/apollo3/mockserver/MockServer.kt#L37) for JS) There's also a test [here](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/apollo-mockserver/src/commonTest/kotlin/com/apollographql/apollo3/mockserver/test/EnqueueTest.kt#L52).

Don't remember precisely, but I think the crash was kind of hard to handle consistently between platforms (I think it was JS hanging forever?). To me a `500` is more realistic as in, a real server would not crash your tests. But I also agree that crashing is more explicit/obvious while crafting the tests. Should I revert?

Reply:
> For native it's handled here actually!

👍 Sorry I missed that.

> JS hanging forever

Urg, that's no good...

> a real server would not crash your tests

That's the thing, MockServer is not a real server. By returning 500 if nothing is enqueued, we're letting tests rely on that behaviour and the GraphQL calls are not balanced with the enqueue() anymore. That might or might not make the tests harder to read.

> Should I revert? 

Your call :-). If you decide to keep the 500 default behaviour, this should be in the release notes somewhere.

Reply:
All right I'll let the fate decide 😅 : if I can make it crash consistently I'll do that.

But in any case I agree with you that the test should be explicitly enquing an error if it expects it later, instead of implicitly relying on this behavior! I'll update that.